### PR TITLE
Vending machines are more dangerous when throwing items

### PR DESCRIPTION
**Pull request by coiax**
:cl: coiax
add: Various vending machines, when shooting their inventory at nearby
people, will "demonstrate their products features". This means that if a
cigarette vending machine throws a lighter at you, it will be on.
Vending machines also choose random products when throwing, rather than
the first available one.
/:cl:

- Lighter light is now flame coloured, and have been minorly refactored
- Secvendors prime flashbangs, and turn on their flashights when
throwing them at people.
- Cigarette vendors light their lighters
- The selection process shuffles the item list before iterating, so it
won't just throw the first thing
- Grenades now use addtimer
- A thrown item now ignites mobs if it is hot.

- [x] Flashbangs explode immediately, and should instead be actually primed
- [x] Lighters thrown at people do not set them on fire

### CODE REVIEW

**Comment by banthebantz** [comment]
funny

**Comment by Bawhoppen** [comment]
that is pretty funny, though it was also nice to have a nice harmless feature before

File: code/game/machinery/vending.dm [hunk 1]
```diff
@@ -864,6 +868,11 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
 	/obj/item/weapon/storage/fancy/cigarettes/cigars = 1, /obj/item/weapon/storage/fancy/cigarettes/cigars/havana = 1, /obj/item/weapon/storage/fancy/cigarettes/cigars/cohiba = 1)
 	refill_canister = /obj/item/weapon/vending_refill/cigarette
 
+/obj/machinery/vending/cigarette/pre_throw(obj/item/I)
+	if(istype(I, /obj/item/weapon/lighter))
+		var/obj/item/weapon/lighter/L = I
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Minor nitpick, but
```
var/obj/item/weapon/lighter/L = I
if(istype(L))
```
Has the same performance and is more readable

File: code/game/machinery/vending.dm [hunk 2]
```diff
@@ -919,6 +928,15 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
 	armor = list(melee = 100, bullet = 100, laser = 100, energy = 100, bomb = 0, bio = 0, rad = 0, fire = 100, acid = 50)
 	resistance_flags = FIRE_PROOF
 
+/obj/machinery/vending/security/pre_throw(obj/item/I)
+	if(istype(I, /obj/item/weapon/grenade))
+		var/obj/item/weapon/grenade/G = I
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Ditto

File: code/game/machinery/vending.dm [hunk 3]
```diff
@@ -919,6 +928,15 @@ IF YOU MODIFY THE PRODUCTS LIST OF A MACHINE, MAKE SURE TO UPDATE ITS RESUPPLY C
 	armor = list(melee = 100, bullet = 100, laser = 100, energy = 100, bomb = 0, bio = 0, rad = 0, fire = 100, acid = 50)
 	resistance_flags = FIRE_PROOF
 
+/obj/machinery/vending/security/pre_throw(obj/item/I)
+	if(istype(I, /obj/item/weapon/grenade))
+		var/obj/item/weapon/grenade/G = I
+		G.prime()
+	else if(istype(I, /obj/item/device/flashlight))
+		var/obj/item/device/flashlight/F = I
```
Comment (Cyberboss) [inline code/game/machinery/vending.dm]: Ditto

File: code/game/objects/items/weapons/cigs_lighters.dm [hunk 4]
```diff
@@ -472,8 +472,10 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 	flags = CONDUCT
 	slot_flags = SLOT_BELT
 	var/lit = 0
+	var/fancy = TRUE
 	heat = 1500
 	resistance_flags = FIRE_PROOF
+	light_color = LIGHT_COLOR_FIRE
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/cigs_lighters.dm]: Unrelated

File: code/game/objects/items/weapons/cigs_lighters.dm [hunk 5]
```diff
@@ -484,37 +486,28 @@ CIGARETTE PACKETS ARE IN FANCY.DM
...
-	. = ..()
-	add_atom_colour(color2hex(randomColor(1)), FIXED_COLOUR_PRIORITY)
-	update_icon()
-
-/obj/item/weapon/lighter/greyscale/update_icon()
-	cut_overlays()
-	var/mutable_appearance/base_overlay = mutable_appearance(icon,"[initial(icon_state)]_base")
-	base_overlay.appearance_flags = RESET_COLOR //the edging doesn't change color
+/obj/item/weapon/lighter/proc/set_lit(new_lit)
+	lit = new_lit
```
Comment (Cyberboss) [inline code/game/objects/items/weapons/cigs_lighters.dm]: before this line
```
if(lit == new_lit) return
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/game/machinery/vending.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/machinery/vending.dm]: I disagree.
I hate casting something to something it's potentially not.

And the standard in the codebase is what's currently here, not your suggestion.

**Comment by kevinz000** [comment]
>vending machine throwing primed flashbang

YES
YES
YES
YEs
NOW CODE A MOBILE VERSION I CAN USE FOR RIOT CONTROL

**Comment by Anonmare** [comment]
When a sec-vend throws a pair of handcuffs at someone, does it cuff them?

**Comment by GunHog** [comment]
> NOW CODE A MOBILE VERSION I CAN USE FOR RIOT CONTROL
wrench_png3129

**Comment by Bawhoppen** [comment]
>When a sec-vend throws a pair of handcuffs at someone, does it cuff them?

add a chance for normal thrown cuffs to cuff a person (like 5% chance)

**Comment by Jalleo** [comment]
Oh god hacked booze vendor by three hacked cig machines. Imagine the chaos.

**Comment by lordpidey** [comment]
Do syringes inject contents upon hitting a carbon mob?

**Comment by octareenroon91** [comment]
Could this maybe just do attack_self before throwing?

**Comment by Cyberboss** [comment]
Terrible idea, octa91

File: code/game/machinery/vending.dm (hunk 1 above)
Comment (duncathan) [inline code/game/machinery/vending.dm]: really there's no standard but i do not think it's worth enforcing at all

File: code/game/objects/items/weapons/cigs_lighters.dm (hunk 5 above)
Comment (duncathan) [inline code/game/objects/items/weapons/cigs_lighters.dm]: meh, that case never gets called as it is so nbd IMO
